From 383cd7226a4f580456b3ce93787cc96eb8cb9e99 Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@freescale.com>
Date: Tue, 21 Jul 2015 14:41:59 +0800
Subject: [PATCH 0156/5242] MLK-11269-2 usb: chipidea: host: avoid check
 PORT_SUSPEND wrongly

commit  9888f842158f4951d2d25f985f394c4266b862b9 from
https://source.codeaurora.org/external/imx/linux-imx.git

At some situations, the ehci_bus_suspend may not set PORT_SUSPEND
due to port is not enabled, so add flag ehci->bus_suspended to
check if ehci_bus_suspend set PORT_SUSPEND or not.

We see "ci_hdrc ci_hdrc.0: timeout waiting for SUSPEND" wrongly when
plug in an unsupported usb device, in that case, the PORT_PE is cleared
and bus_suspend is called.

Signed-off-by: Peter Chen <peter.chen@freescale.com>
(cherry picked from commit 82cfe6c31ee55e5aab1a057a5aaf853b7bb6cd07)
(cherry picked from commit a4fe7e55e84944962d6a7be7bd404865d376811e)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/host.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/chipidea/host.c b/drivers/usb/chipidea/host.c
index 78bad97..cdce7e8 100644
--- a/drivers/usb/chipidea/host.c
+++ b/drivers/usb/chipidea/host.c
@@ -521,7 +521,8 @@ static int ci_ehci_bus_suspend(struct usb_hcd *hcd)
 			 * a delay in suspending the port. Poll until the
 			 * port is suspended.
 			 */
-			if (ehci_handshake(ehci, reg, PORT_SUSPEND,
+			if (test_bit(port, &ehci->bus_suspended) &&
+					ehci_handshake(ehci, reg, PORT_SUSPEND,
 							PORT_SUSPEND, 5000))
 				ehci_err(ehci, "timeout waiting for SUSPEND\n");
 
-- 
1.7.9.5

